util/check_dif_statuses.py

```diff
@@ -0,0 +1,307 @@
+    headers = [
+        "IP", "DIF Updated", "HW Updated", "DIF Author*", "HW Author*",
+        "Functions\nDefined.", "Functions\nImplemented", "Stage"
+    ]
+    for dif_status in dif_statuses:
+        # Color code last modified dates.
+        if dif_status.hw_last_modified > dif_status.dif_last_modified:
+            hw_last_modified = colored(dif_status.hw_last_modified.date(),
+                                       "yellow")
+            dif_last_modified = colored(dif_status.dif_last_modified.date(),
+                                        "yellow")
+        else:
+            hw_last_modified = dif_status.hw_last_modified.date()
+            dif_last_modified = dif_status.dif_last_modified.date()
+        # Color code API complete status.
+        if dif_status.api_complete:
+            num_funcs_defined = colored(dif_status.num_functions_defined,
+                                        "green")
+            num_funcs_implemented = colored(
+                dif_status.num_functions_implemented, "green")
+        else:
+            num_funcs_defined = colored(dif_status.num_functions_defined,
+                                        "red")
+            num_funcs_implemented = colored(
+                dif_status.num_functions_implemented, "red")
+
+        # Add row to table (printing one author email per line).
+        rows.append([
+            dif_status.ip, dif_last_modified, hw_last_modified,
+            "\n".join(dif_status.dif_main_authors),
+            "\n".join(dif_status.hw_main_authors), num_funcs_defined,
+            num_funcs_implemented, dif_status.lifecycle_state
+        ])
+
+    # Print the table and legend.
+    print("DIF Statuses:")
+    print(tabulate(rows, headers, tablefmt=TABLE_FORMAT))
+    print("*Only the lead author (by LOC) for each component is listed.")
+    print(colored("Yellow", "yellow"),
+          "\t= HW has been updated since the DIF.")
+    print(colored("Green", "green"), "\t= DIF API is complete.")
+    print(colored("Red", "red"), "\t= DIF API is incomplete.")
+
+
+def print_unimplemented_functions(dif_statuses):
+    """Print a table of unimplemented DIF functions to STDOUT."""
+    rows = []
+    headers = ["IP", "Function"]
+    for dif_status in dif_statuses:
+        if not dif_status.api_complete:
+            rows.append(
+                [dif_status.ip, "\n".join(dif_status.funcs_unimplemented)])
+    print("Unimplemented Functions:")
+    print(tabulate(rows, headers, tablefmt=TABLE_FORMAT))
+
+
+def main(argv):
+    # Process args.
+    parser = argparse.ArgumentParser(
+        prog="check_dif_statuses",
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument(
+        "repo_top",
+        help=
+        """Relative path to where the OpenTitan repository is checked out.""")
+    parser.add_argument(
+        "--show-unimplemented",
+        action='store_true',
+        help="""Show unimplemented functions for each incomplete DIF.""")
```

Comment:
Just to stay honest about this: The notion of completeness of a DIF based on implemented functions is very naive.  It assumes the header file has a complete description of the API for that DIF and assumes that any implemented function that exists is itself complete.

I think this is a useful estimate; we should just make sure that we're up-front about how the estimate is calculated and what the assumptions are.

Having said all that, I don't exactly know where the right place to put this bit of explanation might be.  It's clearly too long for argument help text :)

Reply:
Good point, best to clearly state this. Maybe I could add a short description in the legend beneath the table definition, next to the "Green = ..." and "Red = ..." (for reference see the updated table look below)? E.g.:
"Green = DIF API, as defined in the current header file, is complete. Note: the header file may lack necessary API functionality."
"Red = DIF API is incomplete, as defined in the header file."

Lmk what you think.

<img width="1164" alt="Screen Shot 2021-07-28 at 4 08 27 PM" src="https://user-images.githubusercontent.com/5633066/127407577-f14f05b3-72b0-403c-be27-dc089219a466.png">

Reply:
sgtm

Reply:
added the short disclaimers to the table legend